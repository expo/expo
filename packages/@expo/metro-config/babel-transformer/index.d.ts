export * from '../build/babel-transformer';
