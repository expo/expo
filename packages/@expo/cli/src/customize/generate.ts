import path from 'path';
import resolveFrom from 'resolve-from';

import { installAsync } from '../install/installAsync';
import { Log } from '../log';
import { copyAsync } from '../utils/dir';
import { CommandError } from '../utils/errors';
import { DestinationResolutionProps, selectTemplatesAsync, TEMPLATES } from './templates';

export async function queryAndGenerateAsync(
  projectRoot: string,
  {
    files,
    props,
    extras,
  }: {
    files: string[];
    props: DestinationResolutionProps;
    /** Any extra props to pass to the install command. */
    extras: any[];
  }
) {
  const valid = files.filter(
    (file) => !!TEMPLATES.find((template) => template.destination(props) === file)
  );

  if (valid.length !== files.length) {
    const diff = files.filter(
      (file) => !TEMPLATES.find((template) => template.destination(props) === file)
    );
    throw new CommandError(
      `Invalid files: ${diff.join(', ')}. Allowed: ${TEMPLATES.map((template) =>
        template.destination(props)
      ).join(', ')}`
    );
  }

  if (!valid.length) {
    return;
  }
  Log.log(`Generating: ${valid.join(', ')}`);
  return generateAsync(projectRoot, {
    answer: files.map((file) =>
      TEMPLATES.findIndex((template) => template.destination(props) === file)
    ),
    props,
    extras,
  });
}

/** Select templates to generate then generate and install. */
export async function selectAndGenerateAsync(
  projectRoot: string,
  {
    props,
    extras,
  }: {
    props: DestinationResolutionProps;
    /** Any extra props to pass to the install command. */
    extras: any[];
  }
) {
  const answer = await selectTemplatesAsync(projectRoot, props);

  if (!answer?.length) {
    Log.exit('\n\u203A Exiting with no change...', 0);
  }

  await generateAsync(projectRoot, {
    answer,
    props,
    extras,
  });
}

async function generateAsync(
  projectRoot: string,
  {
    answer,
    props,
    extras,
  }: {
    answer: number[];
    props: DestinationResolutionProps;
    /** Any extra props to pass to the install command. */
    extras: any[];
  }
) {
  // Copy files
  await Promise.all(
    answer.map((file) => {
      const template = TEMPLATES[file];
      const projectFilePath = path.resolve(projectRoot, template.destination(props));
      // copy the file from template
      return copyAsync(template.file(projectRoot), projectFilePath, {
        overwrite: true,
        recursive: true,
      });
    })
  );

  // Install dependencies
  const packages = answer
    .map((file) => TEMPLATES[file].dependencies)
    .flat()
    .filter((pkg) => !resolveFrom.silent(projectRoot, pkg));
  if (packages.length) {
    Log.debug('Installing ' + packages.join(', '));
    await installAsync(packages, {}, ['--dev', ...extras]);
  }
}
