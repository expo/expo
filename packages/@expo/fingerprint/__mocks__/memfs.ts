const memfs = jest.requireActual('memfs');
const path = require('path');

/**
 * Further mock memfs `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = memfs.fs.promises.mkdtemp;
memfs.fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  memfs.fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = memfs;
