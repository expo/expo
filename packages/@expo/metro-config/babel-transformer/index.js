module.exports = require('../build/babel-transformer');
