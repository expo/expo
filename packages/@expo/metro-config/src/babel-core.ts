// re-exported because babel/core is hard to mock.
export {
  type PluginItem,
  type TransformOptions,
  transformFromAstSync,
  transformSync,
} from '@babel/core';
