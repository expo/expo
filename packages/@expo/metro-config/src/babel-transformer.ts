/**
 * Copyright (c) 650 Industries (Expo). All rights reserved.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// A fork of the upstream babel-transformer that uses Expo-specific babel defaults
// and adds support for web and Node.js environments via `isServer` on the Babel caller.

// @ts-expect-error
import inlineRequiresPlugin from 'babel-preset-fbjs/plugins/inline-requires';
import type { BabelTransformer, BabelTransformerArgs } from 'metro-babel-transformer';
// @ts-expect-error
import makeHMRConfig from 'metro-react-native-babel-preset/src/configs/hmr';
import assert from 'node:assert';
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import resolveFrom from 'resolve-from';

import {
  parseSync,
  type PluginItem,
  type TransformOptions,
  transformFromAstSync,
} from './babel-core';

const cacheKeyParts = [
  fs.readFileSync(__filename),
  require('babel-preset-fbjs/package.json').version,
];

// TS detection conditions copied from metro-react-native-babel-preset
function isTypeScriptSource(fileName: string): boolean {
  return !!fileName && fileName.endsWith('.ts');
}

function isTSXSource(fileName: string): boolean {
  return !!fileName && fileName.endsWith('.tsx');
}

let babelPresetExpo: string | null | undefined = null;

function getBabelPresetExpo(projectRoot: string): string | null {
  if (babelPresetExpo !== undefined) {
    return babelPresetExpo;
  }

  babelPresetExpo = resolveFrom.silent(projectRoot, 'babel-preset-expo') ?? null;
  return babelPresetExpo;
}

/**
 * Return a memoized function that checks for the existence of a
 * project level .babelrc file, and if it doesn't exist, reads the
 * default RN babelrc file and uses that.
 */
const getBabelRC = (function () {
  let babelRC: TransformOptions | null /*: ?BabelCoreOptions */ = null;

  /* $FlowFixMe[missing-local-annot] The type annotation(s) required by Flow's
   * LTI update could not be added via codemod */
  return function _getBabelRC({
    projectRoot,
    extendsBabelConfigPath,
    ...options
  }: BabelTransformerArgs['options']) {
    if (babelRC != null) {
      return babelRC;
    }

    babelRC = {
      plugins: [],
      extends: extendsBabelConfigPath,
    };

    if (extendsBabelConfigPath) {
      return babelRC;
    }

    // Let's look for a babel config file in the project root.
    let projectBabelRCPath;

    // .babelrc
    if (projectRoot) {
      projectBabelRCPath = path.resolve(projectRoot, '.babelrc');
    }

    if (projectBabelRCPath) {
      // .babelrc.js
      if (!fs.existsSync(projectBabelRCPath)) {
        projectBabelRCPath = path.resolve(projectRoot, '.babelrc.js');
      }

      // babel.config.js
      if (!fs.existsSync(projectBabelRCPath)) {
        projectBabelRCPath = path.resolve(projectRoot, 'babel.config.js');
      }

      // If we found a babel config file, extend our config off of it
      // otherwise the default config will be used
      if (fs.existsSync(projectBabelRCPath)) {
        babelRC.extends = projectBabelRCPath;
      }
    }

    // If a babel config file doesn't exist in the project then
    // the default preset for react-native will be used instead.
    if (!babelRC.extends) {
      const { experimentalImportSupport, ...presetOptions } = options;

      // Convert the options into the format expected by the Expo preset.
      const platformOptions = {
        // @ts-expect-error: This is how Metro works by default
        unstable_transformProfile: presetOptions.unstable_transformProfile,
        // @ts-expect-error: This is how Metro works by default
        withDevTools: presetOptions.withDevTools,
        disableImportExportTransform: experimentalImportSupport,
        dev: presetOptions.dev,
        enableBabelRuntime: presetOptions.enableBabelRuntime,
      };

      babelRC.presets = [
        [
          // NOTE(EvanBacon): Here we use the Expo babel wrapper instead of the default react-native preset.
          require('babel-preset-expo'),
          {
            web: platformOptions,
            native: platformOptions,
            // lazyImports: presetOptions.lazyImportExportTransform,
          },
        ],
      ];
    }

    return babelRC;
  };
})();

/**
 * Given a filename and options, build a Babel
 * config object with the appropriate plugins.
 */
function buildBabelConfig(
  filename: string,
  options: BabelTransformerArgs['options'],
  plugins: PluginItem[] = []
): TransformOptions {
  const babelRC = getBabelRC(options);

  const extraConfig: TransformOptions = {
    babelrc: typeof options.enableBabelRCLookup === 'boolean' ? options.enableBabelRCLookup : true,
    code: false,
    cwd: options.projectRoot,
    filename,
    highlightCode: true,
  };

  let config: TransformOptions = {
    ...babelRC,
    ...extraConfig,
  };

  // Add extra plugins
  const extraPlugins = [];

  if (options.inlineRequires) {
    extraPlugins.push(inlineRequiresPlugin);
  }

  config.plugins = extraPlugins.concat(config.plugins, plugins);

  const withExtraPlugins = config.plugins;

  if (options.dev && options.hot) {
    // Note: this intentionally doesn't include the path separator because
    // I'm not sure which one it should use on Windows, and false positives
    // are unlikely anyway. If you later decide to include the separator,
    // don't forget that the string usually *starts* with "node_modules" so
    // the first one often won't be there.
    const mayContainEditableReactComponents = !filename.includes('node_modules');

    if (mayContainEditableReactComponents) {
      const hmrConfig = makeHMRConfig();
      hmrConfig.plugins = withExtraPlugins.concat(hmrConfig.plugins);
      config = { ...config, ...hmrConfig };
    }
  }

  return {
    ...babelRC,
    ...config,
  };
}

const transform: BabelTransformer['transform'] = ({
  filename,
  options,
  src,
  plugins,
}: BabelTransformerArgs): ReturnType<BabelTransformer['transform']> => {
  const OLD_BABEL_ENV = process.env.BABEL_ENV;
  process.env.BABEL_ENV = options.dev ? 'development' : process.env.BABEL_ENV || 'production';

  // Ensure the default babel preset is Expo.
  options.extendsBabelConfigPath = getBabelPresetExpo(options.projectRoot) ?? undefined;

  try {
    const babelConfig: TransformOptions = {
      // ES modules require sourceType='module' but OSS may not always want that
      sourceType: 'unambiguous',
      ...buildBabelConfig(filename, options, plugins),
      caller: {
        name: 'metro',
        // @ts-expect-error: Custom values passed to the caller.
        bundler: 'metro',
        platform: options.platform,
        // Empower the babel preset to know the env it's bundling for.
        // Metro automatically updates the cache to account for the custom transform options.
        isServer: options.customTransformOptions?.environment === 'node',

        isDev: options.dev,
        // Pass the engine to babel so we can automatically transpile for the correct
        // target environment.
        engine: options.customTransformOptions?.engine,

        // Provide the project root for accurately reading the Expo config.
        projectRoot: options.projectRoot,
      },
      ast: true,

      // NOTE(EvanBacon): We split the parse/transform steps up to accommodate
      // Hermes parsing, but this defaults to cloning the AST which increases
      // the transformation time by a fair amount.
      // You get this behavior by default when using Babel's `transform` method directly.
      cloneInputAst: false,
    };
    const sourceAst =
      isTypeScriptSource(filename) || isTSXSource(filename) || !options.hermesParser
        ? parseSync(src, babelConfig)
        : require('hermes-parser').parse(src, {
            babel: true,
            sourceType: babelConfig.sourceType,
          });

    const result = transformFromAstSync(sourceAst, src, babelConfig);

    // The result from `transformFromAstSync` can be null (if the file is ignored)
    if (!result) {
      // BabelTransformer specifies that the `ast` can never be null but
      // the function returns here. Discovered when typing `BabelNode`.
      return { ast: null };
    }

    assert(result.ast);
    return { ast: result.ast, metadata: result.metadata };
  } finally {
    if (OLD_BABEL_ENV) {
      process.env.BABEL_ENV = OLD_BABEL_ENV;
    }
  }
};

function getCacheKey() {
  const key = crypto.createHash('md5');
  cacheKeyParts.forEach((part) => key.update(part));
  return key.digest('hex');
}

const babelTransformer: BabelTransformer = {
  transform,
  getCacheKey,
};

module.exports = babelTransformer;
