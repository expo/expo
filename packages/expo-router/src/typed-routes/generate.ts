import fs from 'node:fs';
import path from 'path';

import { RouteNode } from '../Route';
import { getRoutes } from '../getRoutes';
import { isTypedRoute, removeSupportedExtensions } from '../matchers';
import { RequireContext } from '../types';

// /[...param1]/ - Match [...param1]
const CATCH_ALL = /\[\.\.\..+?\]/g;
// /[param1] - Match [param1]
const SLUG = /\[.+?\]/g;

export function getTypedRoutesDeclarationFile(ctx: RequireContext) {
  const staticRoutes = new Set<string>();
  const dynamicRoutes = new Set<string>();
  const dynamicRouteContextKeys = new Set<string>();

  walkRouteNode(
    getRoutes(ctx, {
      ignoreEntryPoints: true,
      ignoreRequireErrors: true,
      importMode: 'async',
    }),
    staticRoutes,
    dynamicRoutes,
    dynamicRouteContextKeys
  );

  // If the user has expo-router v3+ installed, we can use the types from the package
  return (
    fs
      .readFileSync(path.join(__dirname, '../../types/expo-router.d.ts'), 'utf-8')
      // Swap from being a namespace to a module
      .replace('declare namespace ExpoRouter {', `declare module "expo-router" {`)
      // Add the route values
      .replace(
        'type StaticRoutes = string;',
        `type StaticRoutes = ${setToUnionType(staticRoutes)};`
      )
      .replace(
        'type DynamicRoutes<T extends string> = string;',
        `type DynamicRoutes<T extends string> = ${setToUnionType(dynamicRoutes)};`
      )
      .replace(
        'type DynamicRouteTemplate = never;',
        `type DynamicRouteTemplate = ${setToUnionType(dynamicRouteContextKeys)};`
      )
  );
}

/**
 * Walks a RouteNode tree and adds the routes to the provided sets
 */
function walkRouteNode(
  routeNode: RouteNode | null,
  staticRoutes: Set<string>,
  dynamicRoutes: Set<string>,
  dynamicRouteContextKeys: Set<string>
) {
  if (!routeNode) return;

  addRouteNode(routeNode, staticRoutes, dynamicRoutes, dynamicRouteContextKeys);

  for (const child of routeNode.children) {
    walkRouteNode(child, staticRoutes, dynamicRoutes, dynamicRouteContextKeys);
  }
}

/**
 * Given a RouteNode, adds the route to the correct sets
 * Modifies the RouteNode.route to be a typed-route string
 */
function addRouteNode(
  routeNode: RouteNode | null,
  staticRoutes: Set<string>,
  dynamicRoutes: Set<string>,
  dynamicRouteContextKeys: Set<string>
) {
  if (!routeNode?.route) return;
  if (!isTypedRoute(routeNode.route)) return;

  const routePath = `/${removeSupportedExtensions(routeNode.route).replace(/\/?index$/, '')}`; // replace /index with /

  if (routeNode.dynamic) {
    dynamicRouteContextKeys.add(routePath);
    dynamicRoutes.add(
      `${routePath
        .replaceAll(CATCH_ALL, '${CatchAllRoutePart<T>}')
        .replaceAll(SLUG, '${SingleRoutePart<T>}')}`
    );
  } else {
    staticRoutes.add(routePath);
  }
}

/**
 * Converts a Set to a TypeScript union type
 */
const setToUnionType = <T>(set: Set<T>) => {
  return set.size > 0 ? [...set].map((s) => `\`${s}\``).join(' | ') : 'never';
};
