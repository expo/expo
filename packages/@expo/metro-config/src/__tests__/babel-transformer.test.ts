import generate from '@babel/generator';
import { vol } from 'memfs';
import type { BabelTransformer } from 'metro-babel-transformer';

import * as babel from '../babel-core';
// eslint-disable-next-line import/namespace
import * as untypedTransformer from '../babel-transformer';

const transformer = untypedTransformer as BabelTransformer;

jest.mock('../babel-core', () => {
  const babel = jest.requireActual('../babel-core');
  return {
    ...babel,
    transformFromAstSync: jest.fn((...props) => babel.transformFromAstSync(...props)),
    transformSync: jest.fn((...props) => babel.transformSync(...props)),
  };
});
const originalWarn = console.warn;

afterEach(() => {
  vol.reset();
  console.warn = originalWarn;
});

it(`passes the environment as isServer to the babel preset`, () => {
  console.warn = jest.fn();
  vol.fromJSON({}, '/');

  const fixture = `import { Platform } from 'react-native';
    
    export default function App() {
        return <div>Hello</div>
    }`;

  const results = transformer.transform({
    filename: 'foo.js',
    options: {
      globalPrefix: '',
      enableBabelRuntime: true,
      enableBabelRCLookup: true,
      dev: true,
      projectRoot: '/',
      hot: true,
      inlineRequires: false,
      minify: false,
      platform: 'ios',
      publicPath: '/',
      customTransformOptions: Object.create({
        environment: 'node',
      }),
    },
    src: fixture,
    plugins: [],
  });

  expect(generate(results.ast).code).toMatchSnapshot();

  expect(babel.transformSync).toBeCalledWith(fixture, {
    ast: true,
    babelrc: true,
    caller: {
      // HERE IS THE MAGIC
      isReactServer: false,
      isServer: true,
      isDev: true,
      bundler: 'metro',
      engine: undefined,
      name: 'metro',
      platform: 'ios',
      baseUrl: '',
      isNodeModule: false,
      isHMREnabled: true,
      preserveEnvVars: undefined,
      projectRoot: expect.any(String),
      routerRoot: 'app',
    },
    cloneInputAst: false,
    code: false,
    cwd: '/',
    extends: undefined,
    filename: 'foo.js',
    highlightCode: true,
    presets: [expect.anything()],
    plugins: [],
    sourceType: 'unambiguous',
  });
});

it(`passes the environment as isReactServer to the babel preset`, () => {
  console.warn = jest.fn();
  vol.fromJSON({}, '/');

  const fixture = `import { Platform } from 'react-native';
    
    export default function App() {
        return <div>Hello</div>
    }`;

  const results = transformer.transform({
    filename: 'foo.js',
    options: {
      globalPrefix: '',
      enableBabelRuntime: true,
      enableBabelRCLookup: true,
      dev: true,
      projectRoot: '/',
      hot: true,
      inlineRequires: false,
      minify: false,
      platform: 'ios',
      publicPath: '/',
      customTransformOptions: Object.create({
        environment: 'react-server',
      }),
    },
    src: fixture,
    plugins: [],
  });

  expect(console.warn).toBeCalledTimes(0);
  expect(generate(results.ast).code).toMatchSnapshot();

  expect(babel.transformSync).toBeCalledWith(fixture, {
    ast: true,
    babelrc: true,
    caller: expect.objectContaining({
      // HERE IS THE MAGIC
      isReactServer: true,
      isServer: true,
      isDev: true,
      bundler: 'metro',
      engine: undefined,
      name: 'metro',
      platform: 'ios',
      baseUrl: '',
      isNodeModule: false,
      isHMREnabled: true,
      preserveEnvVars: undefined,
      projectRoot: expect.any(String),
      routerRoot: 'app',
    }),
    cloneInputAst: false,
    code: false,
    cwd: '/',
    extends: undefined,
    filename: 'foo.js',
    highlightCode: true,
    presets: [expect.anything()],
    plugins: [],
    sourceType: 'unambiguous',
  });
});
