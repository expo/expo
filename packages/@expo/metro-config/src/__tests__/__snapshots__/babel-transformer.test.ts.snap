// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes the environment as isReactServer to the babel preset 1`] = `
"Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = App;
var _reactNative = require("react-native");
var _jsxDevRuntime = require("react/jsx-dev-runtime");
var _jsxFileName = "/foo.js";
function App() {
  return /*#__PURE__*/(0, _jsxDevRuntime.jsxDEV)("div", {
    children: "Hello"
  }, void 0, false, {
    fileName: _jsxFileName,
    lineNumber: 4,
    columnNumber: 16
  }, this);
}"
`;

exports[`passes the environment as isServer to the babel preset 1`] = `
"Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = App;
var _reactNative = require("react-native");
var _jsxDevRuntime = require("react/jsx-dev-runtime");
var _jsxFileName = "/foo.js";
function App() {
  return /*#__PURE__*/(0, _jsxDevRuntime.jsxDEV)("div", {
    children: "Hello"
  }, void 0, false, {
    fileName: _jsxFileName,
    lineNumber: 4,
    columnNumber: 16
  }, this);
}"
`;
